lib/compress/zstd_compress.c

```diff
@@ -2803,18 +2883,714 @@ static ZSTD_blockCompressor ZSTD_selectBlockCompressor(ZSTD_strategy strat, int
     return blockCompressor[extDict!=0][(U32)strat];
 }
 
+/*-*************************************
+*  Long distance matching
+***************************************/
+
+/** ZSTD_ldm_getSmallHash() :
+ *  numBits should be <= 32
+ *  @return : the most significant numBits of value */
+static U32 ZSTD_ldm_getSmallHash(U64 value, U32 numBits)
+{
+    assert(numBits <= 32);
+    return (U32)(value >> (64 - numBits));
+}
+
+/** ZSTD_ldm_getChecksum() :
+ *  numBitsToDiscard should be <= 32
+ *  @return : the next most significant 32 bits after numBitsToDiscard */
+static U32 ZSTD_ldm_getChecksum(U64 hash, U32 numBitsToDiscard)
+{
+    assert(numBitsToDiscard <= 32);
+    return (hash >> (64 - 32 - numBitsToDiscard)) & 0xFFFFFFFF;
+}
+
+/** ZSTD_ldm_getTag() ;
+ *  Giving the hash, returns the most significant LDM_HASH_EVERY_LOG bits
+ *  after (32 + hbits) bits.
+ *
+ *  If there are not enough bits remaining, return the last
+ *  LDM_HASH_EVERY_LOG bits. */
+static U32 ZSTD_ldm_getTag(U64 hash, U32 hbits)
+{
+    if (32 - hbits < LDM_HASH_EVERY_LOG) {
+        return hash & LDM_HASH_MASK;
+    } else {
+        return (hash >> (32 - hbits - LDM_HASH_EVERY_LOG)) & LDM_HASH_MASK;
+    }
+}
+
+/** ZSTD_ldm_getBucket() :
+ *  Returns a pointer to the start of the bucket associated with hash. */
+static ldmEntry_t* ZSTD_ldm_getBucket(ldmEntry_t *table, size_t hash)
+{
+    return table + (hash << LDM_BUCKET_SIZE_LOG);
+}
+
+/** ZSTD_ldm_insertEntry() :
+ *  Insert the entry with corresponding hash into the table */
+static void ZSTD_ldm_insertEntry(ldmEntry_t* table, BYTE* bucketOffsets,
+                                 size_t hash, const ldmEntry_t entry)
+{
+    *(ZSTD_ldm_getBucket(table, hash) + bucketOffsets[hash]) = entry;
+    bucketOffsets[hash]++;
+    bucketOffsets[hash] &= LDM_BUCKET_SIZE - 1;
+}
+
+/** ZSTD_ldm_makeEntryAndInsertByTag() :
+ *
+ *  Gets the small hash, checksum, and tag from the rollingHash.
+ *
+ *  If the tag matches LDM_HASH_MASK, then
+ *  creates an ldmEntry from the offset, and inserts it into the hash table.
+ *
+ *  hBits is the length of the small hash, which is the most significant hBits
+ *  of rollingHash. The checksum is the next 32 most significant bits, followed
+ *  by LDM_HASH_EVERY_LOG bits that make up the tag. */
+static void ZSTD_ldm_makeEntryAndInsertByTag(ldmEntry_t* table,
+                                             BYTE* bucketOffsets,
+                                             U64 rollingHash, U32 hBits,
+                                             U32 offset)
+{
+    U32 tag = ZSTD_ldm_getTag(rollingHash, hBits);
```

Comment:
minor : coding style : 
try to `const` whenever possible,
it's a great hint for reviewers that the value is set for the rest of the scope.